txmongo/collection.py

```diff
@@ -479,3 +611,50 @@ def find_and_modify(self, query=None, update=None, upsert=False, **kwargs):
                 # Should never get here because of allowable_errors
                 raise ValueError("Unexpected Error: %s" % (result,))
         defer.returnValue(result.get("value"))
+
+
+    # Distinct findAndModify utility method is needed because traditional
+    # find_and_modify() accepts `sort` kwarg as dict and passes it to
+    # MongoDB command without conversion. But in find_one_and_*
+    # methods we want to take `filter.sort` instances
+    @defer.inlineCallbacks
+    def _new_find_and_modify(self, filter, projection, sort, upsert=None,
+                             return_document=ReturnDocument.BEFORE, **kwargs):
+        validate_is_mapping("filter", filter)
+        if not isinstance(return_document, bool):
+            raise ValueError("return_document must be ReturnDocument.BEFORE "
+                             "or ReturnDocument.AFTER")
+
+        cmd = SON([("findAndModify", self._collection_name),
+                   ("query", filter),
+                   ("new", return_document)])
+        cmd.update(kwargs)
+
+        if projection is not None:
+            cmd["fields"] = self._normalize_fields_projection(projection)
+
+        if sort is not None:
+            cmd["sort"] = dict(sort["orderby"])
+        if upsert is not None:
+            validate_boolean("upsert", upsert)
+            cmd["upsert"] = upsert
+
+        no_obj_error = "No matching object found"
+
+        result = yield self._database.command(cmd, allowable_errors=[no_obj_error])
+        defer.returnValue(result.get("value"))
+
+    def find_one_and_delete(self, filter, projection=None, sort=None, **kwargs):
```

Comment:
There is a terminology clash here. TxMongo traditionally used `filter` kwarg for specifying sorting, hinting and other modifiers. But all new methods use `filter` argument to specify query and `sort` argument to specify sorting (if applicable). So we need to choose whether to be consistend with the rest of txmongo's method args or to be consistend with PyMongo. For now I choosed second option and named all args just like in PyMongo, but it might be changed if there will be objections.